File: 0206-reverse-linked-list/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    if (head != nil) {
        head,end := reverse(head)
        head.Next = nil
        return end
    }
    return head
}

func reverse(currentNode *ListNode) (*ListNode,*ListNode) {
    var current,endNode *ListNode
    if(currentNode.Next != nil) {
        current,endNode = reverse(currentNode.Next)
        current.Next = currentNode
    } else {
        endNode = currentNode
    }
    return currentNode,endNode
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) {
            return null;
        }
        ListNode result = reverseTheList(head);
        head.next = null;
        return result;
        
    }

    public ListNode reverseTheList(ListNode head) {

        if(head == null || head.next == null) {
            return head;
        }

        ListNode reversedHead = reverseList(head.next);

        head.next.next = head;

        return reversedHead;
        
    }
}